### PR TITLE
Use published location for chisel3 repository, if available

### PR DESCRIPTION
**Pull request by mmjconolly**
Add a mechanism to subscribe to a location for the chisel3 repository.

I was looking to add a wake `here` self-location variable to chisel3, but it's not a sifive repository, so I've presumed that it wouldn't be appropriate to add it there. If it is, let me know and I'll submit a PR there

### CODE REVIEW

File: build.wake [hunk 1]
```diff
@@ -1,3 +1,13 @@
+global topic chisel3Root: String
+def chisel3Root =
+  def paths = subscribe chisel3Root
+  match (head paths)
+    Some path  =
+      if (len paths) != 1 then
+        panic "Multiple published paths for chisel3 root: {format paths}"
+      else
+        path
+    None = "chisel3" # default path in workspace
```
Comment (richardxia) [inline build.wake:10]: I think this would be cleaner as a direct pattern match rather than calling `head` on the list, since this is almost a textbook case of when pattern matching works best.

```suggestion
  match paths
    Nil = "chisel3" # default path in workspace
    path, Nil = path
    _ = panic "Multiple published paths for chisel3 root: {format paths}"
```

**Comment by mmjconolly** [comment]
I'm not a fan of scoring is this situation, as there's only one correct answer for the location of a repository.
We don't have for example 3 copies that we're trying to pick between.

**Comment by richardxia** [comment]
> I'm not a fan of scoring is this situation, as there's only one correct answer for the location of a repository.
> We don't have for example 3 copies that we're trying to pick between.

Right, but it's not about having 3 copies to pick between, but overriding someone else who may have published to the topic. This isn't really a blocking concern, but my experience with all the Wit and Wake repos shows that people can and will do things that you did not expect them to as long as it works in their workspace.

**Comment by mmjconolly** [comment]
My ideal situation would be that `chisel3` accepts a patch for `def chisel3Root = here`, do you think they would be ok with us adding a file for that there?

**Comment by richardxia** [comment]
> My ideal situation would be that `chisel3` accepts a patch for `def chisel3Root = here`, do you think they would be ok with us adding a file for that there?

I don't think so. If I were them, I wouldn't accept it, since it doesn't make sense that they have to add in a file for someone else's build system. This lack of path-relocatability is a shortcoming of the combo of Wit and Wake, so we should fix it/work around it within Wit, Wake, and our build rules.

**Comment by terpstra** [comment]
What causes chisel3 to be located in a different location than "chisel3"? I would have thought that api-chisel3-sifive is "in charge" of where chisel3 gets put.

**Comment by richardxia** [comment]
> What causes chisel3 to be located in a different location than "chisel3"? I would have thought that api-chisel3-sifive is "in charge" of where chisel3 gets put.

If one does not use wit and instead uses something like git submodules, then `chisel3` is not guaranteed to be located directly in the workspace root. api-chisel3-sifive has no control over where chisel3 is placed, since wit places all Git repositories into a flat hierarchy at the wit workspace root.

**Comment by mmjconolly** [comment]
> What causes chisel3 to be located in a different location than "chisel3"? I would have thought that api-chisel3-sifive is "in charge" of where chisel3 gets put.

Here's an example, where we started on the journey of trying to run wake rules in a repository layout that matches what a federation checkout would have expected.

https://github.com/sifive/product-coreip-sifive/pull/638

**Comment by terpstra** [comment]
Let me rephrase my question: who is providing the `publish` that fills in the correct location?

**Comment by mmjconolly** [comment]
> Let me rephrase my question: who is providing the `publish` that fills in the correct location?

Whoever moves the repositories would do so by
```
$ cd workspace
$ echo "publish chisel3Root = $PATH_TO_CHISEL3" > env.wake
```

That said, I've previously discussed with Richard that I don't like where this is going and that I don't think having user-created location variables all over the place is the right way forward

**Comment by mmjconolly** [comment]
> Indeed. I would have thought in your new repository scheme, api-chisel3-sifive would just know that chisel3 was a submodule of it and thus be able to report the location correctly. It seems like api-chisel3-sifive always knows the right answer for chisel3; all it needs to know is whether you are using the 'base'-scheme or the 'wit'-scheme.

It's the transition period that I'm trying to cover. If we can make as many things location independent as possible before a transition, then there's less hardcoded things to discover in a more-stressful flag-day situation.

**Comment by jackkoenig** [comment]
What about the idea we had of `Wit` or the other repository management tool being responsible for generating a workspace file with the paths? (a git submodules approach could be for that file to be committed to the top repo)

**Comment by mmjconolly** [comment]
> What about the idea we had of `Wit` or the other repository management tool being responsible for generating a workspace file with the paths? (a git submodules approach could be for that file to be committed to the top repo)

That could work, but in that case I still wouldn't want the mechanism in this PR as it's ad-hoc.
Perhaps if we write a wake function that can read wit-lock files

**Comment by terpstra** [comment]
I like @jackkoenig 's idea. We could have something like:
```
export def findPackageLocation =
  def tree =
    def witlock =
      def lockfile = source "wit-lock.json"
      require Pass (JObject repos) = parseJSONFile lockfile else Nil
      def extractLocation (Pair repo obj) =
        obj // `name`
        | getJString
        | omap (Pair repo _)
      mapPartial extractLocation repos
    def layout =
      def layoutfile = source "workspace-layout.json"
      require Pass body = parseJSONFile layoutfile else Nil
      # ... process the other file format, whatever it is
      Nil
    def locations =  witlock ++ layout
    listToTree (_.getPairFirst <=>* _.getPairFirst) locations
  def find repo =
    def Pair hits _ = tequal (Pair repo "") tree
    hits | head | omap getPairFirst
  find : String => Option String
```

**Comment by terpstra** [comment]
Then you would obviously just have api-chisel-sifive declare: 'def chiselRoot = findPackageLocation "chisel3" | getOrElse "chisel3"`
